Add dismissible compatibility notice to sidebar when editing Cart and Checkout

opr opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR adds a notice to the inspector when editing Cart and Checkout blocks, and their inner blocks. It also adds functionality to dismiss the notice. I reused the `useCompatibilityNotice` functionality we have but used a different local storage key, thus replacing the old compatibility notice, but ensuring the new notice is visible to users who did dismiss the old one.

This PR will add the `CartCheckoutSidebarCompatibilityNotice` to the `InspectorControls` by creating a higher order component which includes it, and adding a filter to maybe include this on the `editor.BlockEdit` hook.

I also moved `BlockSettings` to a HOC to allow us to render the compatibility notice before it. This resulted in some duplicated code removal from `cart/edit.js` and `checkout/edit.tsx`.

I created the HOCs in `cart-checkout-shared`, but left the actual `CartCheckoutSidebarCompatibilityNotice` in `editor-components`.

I reimplemented the old `DefaultNotice` since @senadir is blocked by a [PR in WooCommerce Core](https://github.com/woocommerce/woocommerce/pull/34234).

<!-- Reference any related issues or PRs here -->

Fixes #6851
Fixes #6852
Supersedes #6855
Contains changes from #6881

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<img width="293" alt="image" src="https://user-images.githubusercontent.com/5656702/182901522-cc520aa8-e203-45c5-9aff-de443ee0efbe.png">

### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

1. Before checking this PR out, build `trunk`, clear local storage and visit the Cart/Checkout blocks. Dismiss the compatibility notices on each page. Then check out this PR.
2. Go to the editor, insert the Checkout block.
3. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
4. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
5. Go to the editor, insert the Cart block.
6. Select it, verify you see the notice. Click through each and every inner block and ensure the notice displays for it.
7. Dismiss the notice, click back through the inner blocks and ensure the notice is gone. Reload the page, ensure the notice is still gone.
8. Clear local storage again.
9. Go back to the Cart and Checkout blocks and ensure the notice is back.
10. Create a new page. Add the Checkout Block. Select it and verify the sidebar contains the default warning:
<img width="266" alt="image" src="https://user-images.githubusercontent.com/5656702/183895778-44d3d079-09a4-4806-9096-e80e2a46a318.png">.
Verify the wording says **checkout**.
11. Add the Cart Block, ensure the same notice appears but ensure it says **cart**.


* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Performance Impact

<!-- Please document any known performance impact (positive or negative) here. If negative, provide some rationale for why this is an okay tradeoff or how this will be addressed. -->

### Changelog

> Add notice to Cart and Checkout blocks' inspector controls which links to the list of compatible plugins.

senadir committed:
Add default page notice

senadir committed:
show notice all inner blocks

senadir committed:
support flow when page isnt saved

senadir committed:
switch from where we get the current post id

senadir committed:
update lock

senadir committed:
fix types

opr committed:
Remove old compatibility notices from Cart and Checkout

opr committed:
Remove old CartCheckoutCompatibilityNotice

opr committed:
Move useCompatibilityNotice to sidebar-compatibility-notice directory

opr committed:
Create sidebar compatibility notice hoc

opr committed:
Add isCartOrCheckoutOrInnerBlock function

opr committed:
Refactor defaultNotice to use new isCartOrCheckoutOrInnerBlock func

opr committed:
Remove BlockSettings from checkout edit and export from checkout-shared

opr committed:
Change so component still renders, it is just hidden with display: none

This is required because when it returns null the component gets skipped from being added to the Slot, then when it does return a component, then it gets rendered at the bottom of the Slot. By ensuring it always renders we can have it at the top all the time.

opr committed:
Set the priorities of the hoc filters so compat notice renders first

opr commented on assets/js/editor-components/sidebar-compatibility-notice/use-compatibility-notice.ts:
```diff
@@ -10,7 +10,7 @@ export const useCompatibilityNotice = (
 ): [ boolean, () => void ] => {
 	const [ dismissedNotices, setDismissedNotices ] = useLocalStorageState(
-		`wc-blocks_dismissed_compatibility_notices`,
+		`wc-blocks_dismissed_sidebar_compatibility_notices`,
```
Done so merchants who have dismissed the current notices will see the new one.

senadir replied:
Dismissing the notice for Checkout doesn't dismiss it for Cart, should we unify both?

opr committed:
Make isCartOrCheckoutInnerBlock a hook

opr committed:
Remove old compatibility notice related tests

opr committed:
Remove BlockSettings from Cart

opr committed:
Remove withDefaultNotice hoc

opr committed:
Include DefaultNotice in compatibility notice

opr committed:
Remove DefaultNotice from Checkout

opr commented on assets/js/editor-components/utils/index.js:
```diff
@@ -217,3 +219,29 @@ export const formatTitle = ( page, pages ) => {
+ * @param {string} clientId The clientId of the block.
+ * @return { { isCheckout: boolean, isCart: boolean } } An object with isCheckout and isCart booleans.
+ */
+export const useIsCartOrCheckoutOrInnerBlock = ( clientId ) => {
```
Maybe we need a better name for this?

senadir replied:
Now that this is used in a single place, do we need to have a hook?

opr replied:
Good point, I thought I used it in the `withBlockSettings` hoc but I didn't. Removed it now.

opr committed:
Rename withSidebarCompatibilityNotice to withSidebarNotices

This is because it includes the sidebar compatibility notice and the default notices

opr committed:
Remove useIsCartOrCheckoutOrInnerBlock hook

opr committed:
Remove compatibility notice code from tests

opr committed:
Revert DefaultNotice back to the old one

senadir commented on assets/js/blocks/cart-checkout-shared/block-settings/index.tsx:
```diff
@@ -0,0 +1,70 @@
+				/>
+			</PanelBody>
+			<CartCheckoutFeedbackPrompt />
+		</InspectorControls>
+	);
+};
+
+const withBlockSettings = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const isCartOrCheckout =
+			props.name === 'woocommerce/checkout' ||
+			props.name === '@woocommerce/cart';
+		return (
+			<>
+				{ isCartOrCheckout && (
+					<InspectorControls>
+						<BlockSettings { ...props } />
+					</InspectorControls>
+				) }
+
+				<BlockEdit { ...props } />
+			</>
+		);
+	},
+	'withBlockSettings'
+);
+
+if (
+	! hasFilter(
+		'editor.BlockEdit',
+		'woocommerce/add/cart-checkout-block-settings'
+	)
+) {
+	addFilter(
+		'editor.BlockEdit',
+		'woocommerce/add/cart-checkout-block-settings',
+		withBlockSettings,
+		10
+	);
+}
```
I don't get this abstraction, what are we trying to do here?

opr replied:
We don't have to duplicate it in cart/edit and checkout/edit. This is kind of a hangover from when we ran into the Slot/Fill bug when `CartCheckoutSidebarCompatibilityNotice` returned null.

I will move `BlockSettings` to `cart-checkout-shared` and remove this hoc. Prevents duplication without the hoc abstraction.

senadir commented on assets/js/blocks/cart-checkout-shared/default-notice/editor.scss:
```diff
@@ -0,0 +1,11 @@
+.wc-block-cart__page-notice {
+	margin: 0;
+	padding-right: 16px;
+	.components-notice__dismiss {
+		min-width: 24px;
+	}
+	svg {
+		width: 16px;
+		height: 16px;
+	}
+}
```
Currently, those styles are meant for the default cart notice (that's fine), however, can we extend them to cover the compatibility notice as well?

The goal was to match `x` sizes here
<img width="282" alt="image" src="https://user-images.githubusercontent.com/6165348/183944134-dd1509a6-1001-49bb-b3d0-d4d6cd02efe8.png">

senadir commented on assets/js/editor-components/sidebar-compatibility-notice/index.tsx:
```diff
@@ -0,0 +1,55 @@
+	);
+};
+
+export const CheckoutSidebarCompatibilityNotice = () => (
+	<CartCheckoutSidebarCompatibilityNotice block={ 'checkout' } />
+);
+export const CartSidebarCompatibilityNotice = () => (
+	<CartCheckoutSidebarCompatibilityNotice block={ 'cart' } />
+);
```
Are those 2 still used?

opr replied:
Good catch. No they are not.

senadir reviewed (approved):
Thank you for working on this and cleaning up tests, much appreciated Thomas! I left some notes that can be addressed and for this to be merged after.

opr committed:
Remove unused components

tarhi-saad commented on assets/js/blocks/cart-checkout-shared/block-settings/index.tsx:
```diff
@@ -0,0 +1,70 @@
+					}
+				/>
+			</PanelBody>
+			<CartCheckoutFeedbackPrompt />
```
@opr, previously, when selecting the `Cart` block, we could see the `feedback` notice. But, for some reason, this doesn't show up anymore!

<img src="https://user-images.githubusercontent.com/14235870/184043271-faf9a571-8fe6-40c5-bc70-42ad67b68c2f.png" alt="image" width="300"/>

EDIT 1: I believe the whole `BlockSettings` isn't added to the `Cart` block (e.g., no `Dark mode` option either).

tarhi-saad reviewed (changes requested):
Thank you for your excellent work on this PR, @opr. I left a comment below!

opr committed:
Remove withBlockSettings HOC and fix TS types

This is an abstraction that is no longer required, we can just include BlockSettings in the Cart and Checkout blocks

opr committed:
Remove CartCheckoutFeedbackPrompt from BlockSettings

It will be included in sidebar-notices instead

opr committed:
Fix TS Types in DefaultNotice

opr committed:
Add BlockSettings to cart and checkout edit

opr commented:
Latest changes:
- Remove `withBlockSettings` HOC and the filter to apply it.
- Move `BlockSettings` to `cart-checkout-shared`
- Add `BlockSettings` back to the cart and checkout edit functions
- Move `CartCheckoutFeedbackPrompt` into the `withSidebarNotices` hoc (so it appears in all inner blocks).

tarhi-saad committed:
Editor: Add feedback box to the Cart & Checkout Inner Blocks (#6881)

* Show "Feedback prompt" for all inner blocks

* Fix the "feedback" notice position for these blocks

The "checkout fields", "checkout billing address" and
"checkout shipping address" have the addressFields option which
gets rerendered and placed at the bottom of the inspector controls.

* Tidy up the address-fields hoc

opr committed:
Use correct block name to check for billing or shipping address

opr committed:
Revert "Editor: Add feedback box to the Cart & Checkout Inner Blocks (#6881)"

This reverts commit 5f3d6cf15ce08d9c303e62ec3e0d0315ead1281f.

opr committed:
Add hack to get feedback prompt to render last

opr committed:
Fix TS errors for context and attributes

opr committed:
Include CartCheckoutFeedbackPrompt in accountcontrols & addresscontrols

opr committed:
Do not include feedback prompt if on an address block or contact info

opr committed:
Remove unused hoc for address fields

opr commented:
@senadir The way I removed the need for `withAddressFields` was instead explicitly include `CartCheckoutFeedbackPrompt` within [`addressFieldControls`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-627969b1cedc2351e3e406a16c31338a6183087c251f82e491278180a2c033d9R104) and [`accountControls`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-627969b1cedc2351e3e406a16c31338a6183087c251f82e491278180a2c033d9L148)  and in the hoc ([`withSidebarNotices`](https://github.com/woocommerce/woocommerce-blocks/pull/6869/files#diff-a8b2ace3f7ee9e317dfd806c1da525369598f085d661190512e97b235103e3aeR33)) I checked if we were in a block that shows `addressFieldControls ` or `accountControls` and if so, did not render the `CartCheckoutFeedbackPrompt` a a second time.

senadir reviewed (approved):
This is testing fine for me
